File: mobile/android/app/src/main/java/com/mahala/MahalaCore.kt
package com.mahala

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Wrapper JNI pour la bibliothèque Rust Mahala
 */
object MahalaCore {
    init {
        System.loadLibrary("mahala")
    }

    // Fonctions natives JNI
    @JvmStatic
    external fun initLightClient(fullNodeUrl: String): Int
    
    @JvmStatic
    external fun createWalletFromBiometric(biometricHashHex: String, output: ByteArray, outputLen: Int): Int
    
    @JvmStatic
    external fun getBalance(): Double
    
    @JvmStatic
    external fun getDailyDu(): Double
    
    @JvmStatic
    external fun checkIfSelectedValidator(): Int
    
    @JvmStatic
    external fun sync(): Int

    /**
     * Créer un wallet depuis un hash biométrique (async)
     */
    suspend fun createWalletFromBiometricAsync(biometricHashHex: String): Result<String> {
        return withContext(Dispatchers.IO) {
            try {
                val output = ByteArray(128) // Taille suffisante pour une adresse hex
                val result = createWalletFromBiometric(biometricHashHex, output, output.size)
                
                if (result == 0) {
                    // Trouver la fin de la string (null terminator)
                    val nullIndex = output.indexOf(0)
                    val addressBytes = if (nullIndex > 0) {
                        output.sliceArray(0 until nullIndex)
                    } else {
                        output
                    }
                    val address = String(addressBytes, Charsets.UTF_8)
                    Result.success(address)
                } else {
                    Result.failure(Exception("Failed to create wallet: error code $result"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Obtenir la balance (async)
     */
    suspend fun getBalanceAsync(): Result<Double> {
        return withContext(Dispatchers.IO) {
            try {
                val balance = getBalance()
                Result.success(balance)
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Obtenir le DU du jour (async)
     */
    suspend fun getDailyDuAsync(): Result<Double> {
        return withContext(Dispatchers.IO) {
            try {
                val du = getDailyDu()
                Result.success(du)
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Vérifier si sélectionné comme validateur (async)
     */
    suspend fun checkIfSelectedValidatorAsync(): Result<Boolean> {
        return withContext(Dispatchers.IO) {
            try {
                val selected = checkIfSelectedValidator() == 1
                Result.success(selected)
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Synchroniser avec le réseau (async)
     */
    suspend fun syncAsync(): Result<Unit> {
        return withContext(Dispatchers.IO) {
            try {
                val result = sync()
                if (result == 0) {
                    Result.success(Unit)
                } else {
                    Result.failure(Exception("Sync failed with error code $result"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }
}

File: mobile/ios/Mahala/MahalaCore.swift
//
//  MahalaCore.swift
//  Mahala
//
//  Wrapper Swift pour la bibliothèque Rust Mahala
//

import Foundation

@_silgen_name("init_light_client")
func init_light_client(_ url: UnsafePointer<CChar>) -> Int32

@_silgen_name("create_wallet_from_biometric")
func create_wallet_from_biometric(
    _ hash: UnsafePointer<CChar>,
    _ output: UnsafeMutablePointer<CChar>,
    _ outputLen: Int32
) -> Int32

@_silgen_name("get_balance")
func get_balance() -> Double

@_silgen_name("get_daily_du")
func get_daily_du() -> Double

@_silgen_name("check_if_selected_validator")
func check_if_selected_validator() -> Int32

@_silgen_name("sync")
func sync() -> Int32

/// Wrapper Swift pour Mahala Core
class MahalaCore {
    
    /// Initialiser le light client
    static func initLightClient(fullNodeUrl: String) -> Bool {
        let urlCString = fullNodeUrl.cString(using: .utf8)!
        let result = init_light_client(urlCString)
        return result == 0
    }
    
    /// Créer un wallet depuis un hash biométrique
    static func createWalletFromBiometric(biometricHashHex: String) -> Result<String, Error> {
        let hashCString = biometricHashHex.cString(using: .utf8)!
        let output = UnsafeMutablePointer<CChar>.allocate(capacity: 128)
        defer { output.deallocate() }
        
        let result = create_wallet_from_biometric(hashCString, output, 128)
        
        if result == 0 {
            let address = String(cString: output)
            return .success(address)
        } else {
            return .failure(NSError(
                domain: "MahalaCore",
                code: Int(result),
                userInfo: [NSLocalizedDescriptionKey: "Failed to create wallet: error code \(result)"]
            ))
        }
    }
    
    /// Obtenir la balance
    static func getBalance() -> Double {
        return get_balance()
    }
    
    /// Obtenir le Dividende Universel du jour
    static func getDailyDu() -> Double {
        return get_daily_du()
    }
    
    /// Vérifier si sélectionné comme validateur
    static func checkIfSelectedValidator() -> Bool {
        return check_if_selected_validator() == 1
    }
    
    /// Synchroniser avec le réseau
    static func sync() -> Bool {
        return sync() == 0
    }
    
    // Versions async avec Task
    
    /// Créer un wallet (async)
    static func createWalletFromBiometricAsync(biometricHashHex: String) async throws -> String {
        return try await withCheckedThrowingContinuation { continuation in
            Task.detached {
                let result = createWalletFromBiometric(biometricHashHex: biometricHashHex)
                continuation.resume(with: result)
            }
        }
    }
    
    /// Obtenir la balance (async)
    static func getBalanceAsync() async -> Double {
        return await Task.detached {
            getBalance()
        }.value
    }
    
    /// Obtenir le DU (async)
    static func getDailyDuAsync() async -> Double {
        return await Task.detached {
            getDailyDu()
        }.value
    }
    
    /// Vérifier si sélectionné (async)
    static func checkIfSelectedValidatorAsync() async -> Bool {
        return await Task.detached {
            checkIfSelectedValidator()
        }.value
    }
    
    /// Synchroniser (async)
    static func syncAsync() async -> Bool {
        return await Task.detached {
            sync()
        }.value
    }
}

